core/predicate_check.go

```diff
@@ -36,25 +36,30 @@ func CheckPredicates(rules params.Rules, predicateContext *precompileconfig.Pred
 	}
 
 	// Prepare the predicate storage slots from the transaction's access list
-	predicateArguments := predicate.PreparePredicateStorageSlots(rules, tx.AccessList())
-
-	// If there are no predicates to verify, return early and skip requiring the proposervm block
-	// context to be populated.
-	if len(predicateArguments) == 0 {
-		return predicateResults, nil
-	}
-
-	if predicateContext == nil || predicateContext.ProposerVMBlockCtx == nil {
-		return nil, ErrMissingPredicateContext
-	}
-
-	for address, predicates := range predicateArguments {
-		// Since [address] is only added to [predicateArguments] when there's a valid predicate in the ruleset
-		// there's no need to check if the predicate exists here.
-		predicaterContract := rules.Predicaters[address]
-		res := predicaterContract.VerifyPredicate(predicateContext, predicates)
-		log.Debug("predicate verify", "tx", tx.Hash(), "address", address, "res", res)
-		predicateResults[address] = res
+	for _, el := range tx.AccessList() {
+		predicaterContract, exists := rules.Predicaters[el.Address]
+		if !exists {
+			continue
+		}
+		if predicateContext == nil || predicateContext.ProposerVMBlockCtx == nil {
```

Comment:
well seems that we need to return an empty result in case there is no predicate txs. moving this above would return an error even if there is no predicate in access list.